### PR TITLE
Add new Picker components

### PR DESCRIPTION
**Pull request by bryancunningham-okta**
[DES-5646](https://oktainc.atlassian.net/browse/DES-5646)

## Summary
- Adds 2 new 'picker' components for multiple use case. `Picker` and `PickerWithOptionAdornment`

### Picker
image

#### Small adornment
image

#### Large adornment
image

- [ ] I have confirmed this change with my designer and the Odyssey Design Team.

### CODE REVIEW

File: packages/odyssey-react-mui/src/Autocomplete.tsx [hunk 1]
```diff
@@ -13,36 +13,29 @@
...
 } from "react";
-import styled from "@emotion/styled";
-import { VariableSizeList, ListChildComponentProps } from "react-window";
-import { useTranslation } from "react-i18next";
+// import styled from "@emotion/styled";
+// import { VariableSizeList, ListChildComponentProps } from "react-window";
+// import { useTranslation } from "react-i18next";
 
-import { Field } from "./Field";
+// import { Field } from "./Field";
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/Autocomplete.tsx:35]: Do we want to comment these out?

File: packages/odyssey-react-mui/src/Autocomplete.tsx [hunk 2]
```diff
@@ -246,11 +236,6 @@ export type AutocompleteProps<
 > &
   Pick<HtmlProps, "ariaDescribedBy" | "testId" | "translate">;
 
-const ListboxContainer = styled.div`
-  width: 100%;
-  height: 100%;
-`;
-
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/Autocomplete.tsx:253]: Huh, why was this here?

Comment (bryancunningham-okta) [reply packages/odyssey-react-mui/src/Autocomplete.tsx:253]: This was for virtualization. Got moved to the hook

File: packages/odyssey-react-mui/src/useAutocomplete.tsx [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+        OptionType,
+        HasMultipleChoices,
+        undefined,
+        IsCustomValueAllowed
+      >["onChange"]
+    >
+  >(
+    (event, value, reason, details) => {
+      onChangeProp?.(event, value, reason, details);
+    },
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/useAutocomplete.tsx:165]: At this point, wouldn't it make sense to do `...args` instead?

```suggestion
    (...args) => {
      onChangeProp?.(...args);
    },
```

File: packages/odyssey-react-mui/src/useAutocomplete.tsx [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+
+    const sizeMap = useRef<number[]>([]);
+
+    const getListBoxHeight = useCallback(() => {
+      // 8px of padding top/bottom applied by MUI
+      const COMBINED_LISTBOX_PADDING = 16;
+
+      if (itemData.length > OVERSCAN_ROW_COUNT) {
+        // has a max-height of 40vh set in CSS. This is only set because height needs to be a number
+        return 99999;
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/useAutocomplete.tsx:268]: Can we use `Infinity` instead of `99999`?

```suggestion
      if (itemData.length > OVERSCAN_ROW_COUNT) {
        // has a max-height of 40vh set in CSS. This is only set because height needs to be a number
        return Infinity;
```

File: packages/odyssey-react-mui/src/useAutocomplete.tsx [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+    HTMLDivElement,
+    HTMLAttributes<HTMLElement>
+  >(function (props, ref) {
+    const [listHeight, setListHeight] = useState(0);
+
+    const { children, ...other } = props;
+    const itemData: ReactElement[] = (children as ReactElement[]).flatMap(
+      (item: ReactElement & { children?: ReactElement[] }) =>
+        [item].concat(item.children || []),
+    );
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/useAutocomplete.tsx:258]: We shouldn't need to do `children as ReactElement[]`.

If TypeScript isn't seeing it, that's because you need to do `Children.toArray(children).flatMap` instead.

File: packages/odyssey-react-mui/src/useAutocomplete.tsx [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+
+  const useResetCache = (length: number) => {
+    const resetCacheRef = useRef<VariableSizeList>(null);
+    useEffect(() => {
+      if (resetCacheRef.current) {
+        resetCacheRef.current.resetAfterIndex(0, true);
+      }
+    }, [length]);
+    return resetCacheRef;
+  };
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/useAutocomplete.tsx:246]: What's `useResetCache` doing for us? Is there an `X` somewhere that this resets?

Why store the data outside of hooks like `useState`?

File: packages/odyssey-react-mui/src/useAutocomplete.tsx [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+        return 99999;
+      } else {
+        const itemsHeightCalculated = itemData
+          .map((_, index) => sizeMap.current[index] || 0)
+          .reduce(
+            (prevItemHeight, nextItemHeight) => prevItemHeight + nextItemHeight,
+            0,
+          );
+        return COMBINED_LISTBOX_PADDING + itemsHeightCalculated;
+      }
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/useAutocomplete.tsx:277]: This code is confusing because it looks like we're looking `itemData`, but in reality, all we care about are the item indexes. Since they're sequential, we can create a new array with that many items.

```suggestion
        const itemsHeightCalculated = Array(itemData.length).fill(null)
          .map((_, index) => sizeMap.current[index] || 0)
          .reduce(
            (prevItemHeight, nextItemHeight) => prevItemHeight + nextItemHeight,
            0,
          );
        return COMBINED_LISTBOX_PADDING + itemsHeightCalculated;
      }
```

But even better, we can use `slice`:

```suggestion
        const itemsHeightCalculated = sizeMap.current
          .slice(0, itemData.length - 1)
          .reduce(
            (prevItemHeight, nextItemHeight = 0) => prevItemHeight + nextItemHeight,
            0,
          );
        return COMBINED_LISTBOX_PADDING + itemsHeightCalculated;
      }
```

---

Also, if `sizeMap` is a ref, we should call it `sizeMapRef`. Having a random `.current` out there is confusing. My assumption was that `sizeMap` is a `new Map()`.

File: packages/odyssey-react-mui/src/useAutocomplete.tsx (hunk 5 above)
Comment (bryancunningham-okta) [reply packages/odyssey-react-mui/src/useAutocomplete.tsx:258]: I just moved this code. Not sure what the original reasoning was

File: packages/odyssey-react-mui/src/useAutocomplete.tsx [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+      return [] as AutocompleteValue<
+        OptionType,
+        HasMultipleChoices,
+        undefined,
+        IsCustomValueAllowed
+      >;
+    }
+    return value === undefined ? defaultValue : undefined;
+  }, [defaultValue, hasMultipleChoices, value]);
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/useAutocomplete.tsx:138]: I thought all this `hasMultipleChoices` logic was in MUI. If it's something we have to check for all over the place, this is why I suggest we separate them into separate components.

Is this the only place it's relevant?

File: packages/odyssey-react-mui/src/useAutocomplete.tsx (hunk 3 above)
Comment (bryancunningham-okta) [reply packages/odyssey-react-mui/src/useAutocomplete.tsx:165]: I don't have strong feelings either way

File: packages/odyssey-react-mui/src/useAutocomplete.tsx (hunk 4 above)
Comment (bryancunningham-okta) [reply packages/odyssey-react-mui/src/useAutocomplete.tsx:268]: I can't use `Infinity` here. It's applying this number to a height under the hood and `Infinity` is invalid in that scenario

File: packages/odyssey-react-mui/src/useAutocomplete.tsx (hunk 6 above)
Comment (bryancunningham-okta) [reply packages/odyssey-react-mui/src/useAutocomplete.tsx:246]: Something to do with virtualization. I just moved this code from `Autocomplete` didn't look too deeply into ti

File: packages/odyssey-react-mui/src/useAutocomplete.tsx (hunk 8 above)
Comment (bryancunningham-okta) [reply packages/odyssey-react-mui/src/useAutocomplete.tsx:133]: I think this is the only place we do any type of checking. But, I also feel like this could be a simpler. Again, I just moved this code into the hook for reusability.

File: packages/odyssey-react-mui/src/Autocomplete.tsx [hunk 9]
```diff
@@ -286,287 +255,35 @@ const Autocomplete = <
...
+  const {
+    inputValueProp,
+    isVirtualized,
+    onChange,
+    onInputChange,
+    renderInput,
+    t,
+    valueProps,
+    ListboxComponent,
+  } = useAutocomplete<OptionType, HasMultipleChoices, IsCustomValueAllowed>({
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/Autocomplete.tsx:267]: This shouldn't return `t` right? If we need the translation function, we should grab it from that `useTranslation` hook right?

File: packages/odyssey-react-mui/src/Autocomplete.tsx [hunk 10]
```diff
@@ -286,287 +255,35 @@ const Autocomplete = <
...
-    );
+  const {
+    inputValueProp,
+    isVirtualized,
+    onChange,
+    onInputChange,
+    renderInput,
+    t,
+    valueProps,
+    ListboxComponent,
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/Autocomplete.tsx:266]: Interesting, the whole `ListboxComponent` comes from the hook. Never seen that before, but I think it makes sense here.

File: packages/odyssey-react-mui/src/labs/OdysseyPickers/ComposablePicker.tsx [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+  testId,
+  translate,
+}: ComposablePickerProps<
+  OptionType,
+  HasMultipleChoices,
+  IsCustomValueAllowed
+>) => {
+  const {
+    inputValueProp,
+    isVirtualized,
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/labs/OdysseyPickers/ComposablePicker.tsx:124]: If this is a ref, we should name it `isVirtualizedRef` just so it's clear it has a `.current`.

File: packages/odyssey-react-mui/src/labs/OdysseyPickers/ComposablePicker.tsx [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+      translate={translate}
+      open={true}
+    />
+  );
+};
+
+// Need the `typeof Autocomplete` because generics don't get passed through
+const MemoizedComposablePicker = memo(
+  ComposablePicker,
+) as typeof ComposablePicker;
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/labs/OdysseyPickers/ComposablePicker.tsx:187]:
```suggestion
// Need `typeof ComposablePicker` because generics don't get passed through
const MemoizedComposablePicker = memo(
  ComposablePicker,
) as typeof ComposablePicker;
```

File: packages/odyssey-react-mui/src/labs/OdysseyPickers/PickerWithOptionAdornment.tsx [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+  const isImageAdornment = typeof adornment === "string";
+
+  if (isImageAdornment) {
+    return (
+      <OptionAdornmentContainer
+        adornmentSize={adornmentSize}
+        odysseyDesignTokens={odysseyDesignTokens}
+      >
+        {/* NOTE: Intentionally leaving alt as an empty string here so screen readers will ignore this image */}
+        {/* Image should be suffciently described by the adjacent title and/or description of the option */}
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/labs/OdysseyPickers/PickerWithOptionAdornment.tsx:140]:
```suggestion
        {/* Image should be sufficiently described by the adjacent title and/or description of the option */}
```

Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/src/labs/OdysseyPickers/PickerWithOptionAdornment.tsx:140]: The [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) library helps out with these.

File: packages/odyssey-react-mui/src/labs/OdysseyPickers/PickerWithOptionAdornment.tsx [hunk 14]
```diff
@@ -0,0 +1,430 @@
...
+      </OptionAdornmentContainer>
+    );
+  } else {
+    return (
+      <OptionAdornmentContainer
+        adornmentSize={adornmentSize}
+        odysseyDesignTokens={odysseyDesignTokens}
+      >
+        <>{adornment}</>
+      </OptionAdornmentContainer>
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/labs/OdysseyPickers/PickerWithOptionAdornment.tsx:151]: We could just do this right?
```suggestion
      <OptionAdornmentContainer
        adornmentSize={adornmentSize}
        odysseyDesignTokens={odysseyDesignTokens}
      >
        {adornment}
      </OptionAdornmentContainer>
```

If not, we should change to `ReactNode` instead of `ReactElement`.

File: packages/odyssey-react-mui/src/Autocomplete.tsx (hunk 9 above)
Comment (bryancunningham-okta) [reply packages/odyssey-react-mui/src/Autocomplete.tsx:267]: @KevinGhadyani-Okta I had mixed feelings about this. Felt a little weird to return it here but, I also felt like it might make sense to do it to keep from calling the hook again. I'm good either way

File: packages/odyssey-react-mui/src/Autocomplete.tsx (hunk 10 above)
Comment (bryancunningham-okta) [reply packages/odyssey-react-mui/src/Autocomplete.tsx:266]: Yeah, I _could_ just break this out into another component if we want to. But, also felt like it made sense to return it from the hook also

File: packages/odyssey-react-mui/src/useAutocomplete.tsx (hunk 3 above)
Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/src/useAutocomplete.tsx:140]: I'd like to do that then.

File: packages/odyssey-react-mui/src/useAutocomplete.tsx [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+        HasMultipleChoices,
+        undefined,
+        IsCustomValueAllowed
+      >["onInputChange"]
+    >
+  >(
+    (event, value, reason) => {
+      onInputChangeProp?.(event, value, reason);
+    },
+    [onInputChangeProp],
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/useAutocomplete.tsx:157]:
```suggestion
    (...args) => {
      onInputChangeProp?.(...args);
    },
    [onInputChangeProp],
```

File: packages/odyssey-react-mui/src/useAutocomplete.tsx (hunk 5 above)
Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/src/useAutocomplete.tsx]: Ok, let's update it to use `Children.toArray` :). That way we're not masking potential "code breaks something in an app" issues with `as`. That's what happened with Side Nav recently; an `as` masked the fact that it would break when given the wrong value.

Comment (bryancunningham-okta) [reply packages/odyssey-react-mui/src/useAutocomplete.tsx]: I can't get `Children.toArray` to work quite right. TS isn't liking it

**Comment by KevinGhadyani-Okta** [comment]
This PR is almost done! Code's in a great state, and 99% of my comments were on minor updates to improve tests.

**Comment by KevinGhadyani-Okta** [comment]
I would like to see some unit tests, but it's overall fine after the changes we made the other day.

**Review by KevinGhadyani-Okta — APPROVED** [review]

**Review by jordankoschei-okta — APPROVED** [review]
